File: bench/crc32/crc32.go
package main

import (
	"fmt"
	"time"
)

const (
	n       = 16384
	nIters  = 220
	nRuns   = 21
	nWarmup = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

func buildTable(table []uint32) {
	for n := uint32(0); n < 256; n++ {
		c := n
		for k := 0; k < 8; k++ {
			if c&1 != 0 {
				c = 0xedb88320 ^ (c >> 1)
			} else {
				c >>= 1
			}
		}
		table[n] = c
	}
}

func initBuf(buf []byte) {
	x := uint32(0x12345678)
	for i := range buf {
		x = x*1103515245 + 12345
		buf[i] = byte((x >> 16) & 0xff)
	}
}

func crc32Kernel(buf []byte, table []uint32) uint32 {
	c := uint32(0xffffffff)
	for _, b := range buf {
		c = table[(c^uint32(b))&0xff] ^ (c >> 8)
	}
	return c ^ 0xffffffff
}

func runKernel(buf []byte, table []uint32) uint32 {
	h := uint32(0)
	for it := 0; it < nIters; it++ {
		h = mix(h, crc32Kernel(buf, table))
		j := it % n
		buf[j] = (buf[j] + 1) & 0xff
	}
	return h
}

func main() {
	buf := make([]byte, n)
	table := make([]uint32, 256)
	buildTable(table)
	initBuf(buf)
	cs := uint32(0)
	for i := 0; i < nWarmup; i++ {
		cs = runKernel(buf, table)
	}
	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		cs = runKernel(buf, table)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), cs, n*nIters, 1, nRuns)
}

File: bench/crc32/crc32.rs
use std::time::Instant;

const N: usize = 16384;
const N_ITERS: usize = 220;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn build_table(table: &mut [u32; 256]) {
    for n in 0..256 {
        let mut c = n as u32;
        for _ in 0..8 {
            c = if c & 1 != 0 { 0xedb8_8320 ^ (c >> 1) } else { c >> 1 };
        }
        table[n] = c;
    }
}

fn init_buf(buf: &mut [u8]) {
    let mut x = 0x1234_5678u32;
    for b in buf {
        x = x.wrapping_mul(1103515245).wrapping_add(12345);
        *b = ((x >> 16) & 0xff) as u8;
    }
}

fn crc32_kernel(buf: &[u8], table: &[u32; 256]) -> u32 {
    let mut c = 0xffff_ffffu32;
    for &b in buf {
        c = table[((c ^ b as u32) & 0xff) as usize] ^ (c >> 8);
    }
    c ^ 0xffff_ffff
}

fn run_kernel(buf: &mut [u8], table: &[u32; 256]) -> u32 {
    let mut h = 0u32;
    for it in 0..N_ITERS {
        h = mix(h, crc32_kernel(buf, table));
        let j = it % N;
        buf[j] = buf[j].wrapping_add(1);
    }
    h
}

fn main() {
    let mut buf = vec![0; N];
    let mut table = [0u32; 256];
    build_table(&mut table);
    init_buf(&mut buf);
    let mut cs = 0;
    for _ in 0..N_WARMUP {
        cs = run_kernel(&mut buf, &table);
    }
    let mut samples = [0.0; N_RUNS];
    for s in &mut samples {
        let t0 = Instant::now();
        cs = run_kernel(&mut buf, &table);
        *s = t0.elapsed().as_secs_f64() * 1000.0;
    }
    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples),
        cs,
        N * N_ITERS,
        1,
        N_RUNS
    );
}
